File: frontend/src/services/audit.js
/**
 * Audit and Gemini Usage Tracking service client.
 */

export async function fetchAuditLogs(params = {}) {
  const query = new URLSearchParams();
  if (params.page) query.append('page', params.page);
  if (params.pageSize) query.append('page_size', params.pageSize);
  if (params.search) query.append('search', params.search);
  if (params.intent) query.append('intent', params.intent);
  if (params.status) query.append('status', params.status);
  if (params.cacheHit !== undefined && params.cacheHit !== '') query.append('cache_hit', params.cacheHit);
  if (params.needsHumanReview !== undefined && params.needsHumanReview !== '') query.append('needs_human_review', params.needsHumanReview);

  const qs = query.toString();
  const url = `/api/audit${qs ? `?${qs}` : ''}`;

  const response = await fetch(url);
  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}));
    throw new Error(errorData.detail || `Failed to fetch audit logs: ${response.status}`);
  }
  return await response.json();
}

export async function fetchAuditLogDetail(logId) {
  const response = await fetch(`/api/audit/${logId}`);
  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}));
    throw new Error(errorData.detail || `Failed to fetch audit log detail: ${response.status}`);
  }
  return await response.json();
}

export async function fetchGeminiUsage() {
  const response = await fetch('/api/usage');
  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}));
    throw new Error(errorData.detail || `Failed to fetch usage metrics: ${response.status}`);
  }
  return await response.json();
}

File: backend/routes/audit.py
from typing import Optional
from fastapi import APIRouter, HTTPException, Query

from backend.models.audit import AuditLogsResponse, AuditLogDetail, GeminiUsageResponse
from backend.services.audit_service import AuditService

router = APIRouter(tags=["Audit & Governance"])


@router.get("/api/audit", response_model=AuditLogsResponse)
def get_audit_logs(
    page: int = Query(1, ge=1, description="Page number"),
    page_size: int = Query(20, ge=1, le=100, description="Items per page"),
    search: Optional[str] = Query(None, description="Search query in question"),
    intent: Optional[str] = Query(None, description="Filter by classified intent"),
    status: Optional[str] = Query(None, description="Filter by status (success, fallback, human_review, error)"),
    cache_hit: Optional[bool] = Query(None, description="Filter by cache hit boolean"),
    needs_human_review: Optional[bool] = Query(None, description="Filter by human review requirement"),
):
    """
    Retrieve paginated audit logs for Copilot interactions with comprehensive filtering.
    """
    try:
        return AuditService.get_audit_logs(
            page=page,
            page_size=page_size,
            search=search,
            intent=intent,
            status=status,
            cache_hit=cache_hit,
            needs_human_review=needs_human_review,
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to fetch audit logs: {str(e)}")


@router.get("/api/audit/{log_id}", response_model=AuditLogDetail)
def get_audit_log_detail(log_id: int):
    """
    Retrieve full execution trace and evidence for a specific Copilot audit log.
    """
    log_detail = AuditService.get_audit_log_by_id(log_id)
    if not log_detail:
        raise HTTPException(status_code=404, detail=f"Audit log #{log_id} not found")
    return log_detail


@router.get("/api/usage", response_model=GeminiUsageResponse)
def get_gemini_usage():
    """
    Retrieve aggregate Gemini API telemetry, token consumption, cache performance, and cost transparency.
    """
    try:
        return AuditService.get_usage_metrics()
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to calculate usage metrics: {str(e)}")
